### ISSUE TITLE
Docker build says free(): invalid pointer with valid Dockerfile

### ISSUE THREAD

**Issue by peak-load** [docker/for-linux#563]
* [x] This is a bug report and I believe it's related to #562 

### Expected behavior
No errors or exceptions shown

### Actual behavior
User see errors too much debug info right after start of docker build.

### Steps to reproduce the behavior

**Create Dockerfile**
 
```
FROM debian:stable-slim
LABEL description="Build gluon with Debian"
LABEL maintainer="peak-load"

ENV LANG C.UTF-8
RUN apt update && apt -y upgrade && apt install -y bzip2 gawk g++ gcc git libncurses5-dev libssl-dev make python unzip wget zip zlib1g-dev time sudo && rm -rf /var/lib/apt/lists/*
RUN useradd -m -s /bin/bash -G sudo -p builder builder

WORKDIR /home/builder
ADD . /home/builder/src
USER builder
CMD chown -R builder:builder /home/builder/src
WORKDIR /home/builder/src
```

**Start docker build with Dockerfile above:**

```
$ docker build -t peak-load/gluon-build .
free(): invalid pointer
SIGABRT: abort
PC=0x7faa775a4e97 m=0 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0x4afd50, 0xc42004fcc0, 0xc42004fce8)
	/usr/lib/go-1.8/src/runtime/cgocall.go:131 +0xe2 fp=0xc42004fc90 sp=0xc42004fc50
github.com/docker/docker-credential-helpers/secretservice._Cfunc_free(0x153e270)
	github.com/docker/docker-credential-helpers/secretservice/_obj/_cgo_gotypes.go:111 +0x41 fp=0xc42004fcc0 sp=0xc42004fc90
github.com/docker/docker-credential-helpers/secretservice.Secretservice.List.func5(0x153e270)
	/build/golang-github-docker-docker-credential-helpers-cMhSy1/golang-github-docker-docker-credential-helpers-0.5.0/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/secretservice/secretservice_linux.go:96 +0x60 fp=0xc42004fcf8 sp=0xc42004fcc0
github.com/docker/docker-credential-helpers/secretservice.Secretservice.List(0x0, 0x756060, 0xc420018370)
	/build/golang-github-docker-docker-credential-helpers-cMhSy1/golang-github-docker-docker-credential-helpers-0.5.0/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/secretservice/secretservice_linux.go:97 +0x217 fp=0xc42004fda0 sp=0xc42004fcf8
github.com/docker/docker-credential-helpers/secretservice.(*Secretservice).List(0x77e548, 0xc42004fe88, 0x410022, 0xc4200182c0)
	<autogenerated>:4 +0x46 fp=0xc42004fde0 sp=0xc42004fda0
github.com/docker/docker-credential-helpers/credentials.List(0x756ba0, 0x77e548, 0x7560e0, 0xc42000e018, 0x0, 0x10)
	/build/golang-github-docker-docker-credential-helpers-cMhSy1/golang-github-docker-docker-credential-helpers-0.5.0/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/credentials/credentials.go:145 +0x3e fp=0xc42004fe68 sp=0xc42004fde0
github.com/docker/docker-credential-helpers/credentials.HandleCommand(0x756ba0, 0x77e548, 0x7fff936f9764, 0x4, 0x7560a0, 0xc42000e010, 0x7560e0, 0xc42000e018, 0x40e398, 0x4d35c0)
	/build/golang-github-docker-docker-credential-helpers-cMhSy1/golang-github-docker-docker-credential-helpers-0.5.0/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/credentials/credentials.go:60 +0x16d fp=0xc42004fed8 sp=0xc42004fe68
github.com/docker/docker-credential-helpers/credentials.Serve(0x756ba0, 0x77e548)
	/build/golang-github-docker-docker-credential-helpers-cMhSy1/golang-github-docker-docker-credential-helpers-0.5.0/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/credentials/credentials.go:41 +0x1cb fp=0xc42004ff58 sp=0xc42004fed8
main.main()
	/build/golang-github-docker-docker-credential-helpers-cMhSy1/golang-github-docker-docker-credential-helpers-0.5.0/secretservice/cmd/main_linux.go:9 +0x4f fp=0xc42004ff88 sp=0xc42004ff58
runtime.main()
	/usr/lib/go-1.8/src/runtime/proc.go:185 +0x20a fp=0xc42004ffe0 sp=0xc42004ff88
runtime.goexit()
	/usr/lib/go-1.8/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc42004ffe8 sp=0xc42004ffe0

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/lib/go-1.8/src/runtime/asm_amd64.s:2197 +0x1

rax    0x0
rbx    0x7fff936f90e0
rcx    0x7faa775a4e97
rdx    0x0
rdi    0x2
rsi    0x7fff936f8e70
rbp    0x7fff936f91e0
rsp    0x7fff936f8e70
r8     0x0
r9     0x7fff936f8e70
r10    0x8
r11    0x246
r12    0x7fff936f90e0
r13    0x1000
r14    0x0
r15    0x30
rip    0x7faa775a4e97
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
Sending build context to Docker daemon  3.072kB
Step 1/11 : FROM debian:stable-slim
stable-slim: Pulling from library/debian
2e5698f08c0a: Pull complete 
Digest: sha256:68cd99345058647b36cb8e8c1629fe20d81be86a47b0c80fd32492729fa4fe9b
Status: Downloaded newer image for debian:stable-slim
 ---> b24edbef9d7d
Step 2/11 : LABEL description="Build gluon with Debian"
 ---> Running in 1256086790cf
Removing intermediate container 1256086790cf
 ---> 8f40441f2e73
Step 3/11 : LABEL maintainer="peak-load"
 ---> Running in 739719afc137
Removing intermediate container 739719afc137
 ---> e643001f5b85
Step 4/11 : ENV LANG C.UTF-8
 ---> Running in 544e16df5aae
Removing intermediate container 544e16df5aae
 ---> 2c3dad591fbe
Step 5/11 : RUN apt update && apt -y upgrade && apt install -y bzip2 gawk g++ gcc git libncurses5-dev libssl-dev make python unzip wget zip zlib1g-dev time sudo && rm -rf /var/lib/apt/lists/*
 ---> Running in dd3ba2ba18b4

...

Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
Removing intermediate container dd3ba2ba18b4
 ---> 44946ddc63a0
Step 6/11 : RUN useradd -m -s /bin/bash -G sudo -p builder builder
 ---> Running in 96265ad67cc4
Removing intermediate container 96265ad67cc4
 ---> 1c7f97554595
Step 7/11 : WORKDIR /home/builder
 ---> Running in d2b065af68d0
Removing intermediate container d2b065af68d0
 ---> af1a0e5fbf38
Step 8/11 : ADD . /home/builder/src
 ---> c3fc511134f1
Step 9/11 : USER builder
 ---> Running in 87daa4d35a3e
Removing intermediate container 87daa4d35a3e
 ---> 87db5bed5173
Step 10/11 : CMD chown -R builder:builder /home/builder/src
 ---> Running in bf0459f08aa7
Removing intermediate container bf0459f08aa7
 ---> aab31bf23564
Step 11/11 : WORKDIR /home/builder/src
 ---> Running in af84bb180b0a
Removing intermediate container af84bb180b0a
 ---> 1e075ef29c0b
Successfully built 1e075ef29c0b
Successfully tagged peak-load/gluon-build:latest
```

**Output of `docker version`:**

```
docker version
Client:
 Version:           18.09.1
 API version:       1.39
 Go version:        go1.10.6
 Git commit:        4c52b90
 Built:             Wed Jan  9 19:35:31 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.1
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       4c52b90
  Built:            Wed Jan  9 19:02:44 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
$ docker version
Client:
 Version:           18.09.1
 API version:       1.39
 Go version:        go1.10.6
 Git commit:        4c52b90
 Built:             Wed Jan  9 19:35:31 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.1
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       4c52b90
  Built:            Wed Jan  9 19:02:44 2019
  OS/Arch:          linux/amd64
  Experimental:     false
eject@ubuntu-home:~/gluon$ docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 11
Server Version: 18.09.1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 96ec2177ae841256168fcf76954f7177af9446eb
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-43-generic
Operating System: Ubuntu 18.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.61GiB
Name: ubuntu-home
ID: GWFF:SIL7:C4ON:NYIV:6YZF:P4K7:U5KG:TWFK:U632:F45D:Y5RJ:22E7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**
Physical bare metal server.

**Comment by popbones** [comment]
I got this exact same error with a newly installed docker in a newly created instance running ubuntu 18.04 LTS

**Comment by quvide** [comment]
I have the same issue on a virtualized Ubuntu 18.04

**Comment by pyrat** [comment]
I have the same issue on a virtualized Ubuntu 18.04

**Comment by djonik1562** [comment]
I have the same issue on a virtualized Ubuntu 18.04

**Comment by trav0r** [comment]
Uninstalling golang-docker-credential-helpers resolved this issue for me.

Unfortunately, this workaround doesn't allow the installation of docker-compose on 18.04.

**Comment by fergalmoran** [comment]
Same issue on a fresh 18.10 install.

**Comment by steverukuts** [comment]
I'm seeing the same message on Ubuntu 18.04 running on AWS EC2. Builds completed successfully despite this message but I was unable to log in to a remote repository because I experienced the issues described here: https://github.com/docker/compose/issues/6023

I also needed docker-compose for another build configuration, so I uninstalled the `golang-docker-credential-helpers` package as suggested earlier and then installed docker-compose with the [official instructions](https://docs.docker.com/compose/install/). This resolved the problem.

To conclude, it seems there is some sort of conflict caused by the Ubuntu packages not being compatible with each other and this is not strictly a Docker problem - at least this is the position taken on the docker-compose issue I linked.

I am not 100% sure but I believe this is the specific Ubuntu bug that references this behaviour: https://bugs.launchpad.net/ubuntu/+source/golang-github-docker-docker-credential-helpers/+bug/1813003

**Comment by chojrak11** [comment]
It seems it has already been fixed in docker-credential-helpers v0.6.2. The only problem is that Ubuntu uses 0.5.0.
I think the issue should be closed.

**Comment by vlad88sv** [comment]
Same issue with Ubuntu 9.10 and docker official repo.

docker-credential-secretservice version
0.6.0

**Comment by Corruptsector** [comment]
Thanks Rodrigoreis, your fix worked great and fixed it for me.

**Comment by Snokendrutt** [comment]
Worked for me aswell 👍

**Comment by ddouglas** [comment]
For me as well. Thanks @rodrigoreis

**Comment by orenzamir5** [comment]
Works for me too! 10X

**Comment by thaJeztah** [comment]
Yes, this should be fixed in v0.6.3 of the credential helpers; related to https://github.com/docker/docker-credential-helpers/pull/29 and https://github.com/docker/docker-credential-helpers/commit/87c80bfba583eadc087810d17aa631ef4e405efc

**Comment by darrenleeweber** [comment]
Followed the instructions to download and manually install 0.6.3 and got the following error:

```
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.3 LTS"

$ docker-credential-secretservice version
0.6.3

$ docker-credential-secretservice list
Error from list function in secretservice_linux.c likely due to error in secretservice library
```

**Comment by thaJeztah** [comment]
@darrenleeweber that error looks to be generated here; https://github.com/docker/docker-credential-helpers/blob/1c9f7ede70a5ab9851f4c9cb37d317fd89cd318a/secretservice/secretservice_linux.go#L99

Perhaps you can open a ticket in the https://github.com/docker/docker-credential-helpers issue tracker?

**Comment by darrenleeweber** [comment]
My 2c - uninstall the credential helper and get on with my real work

**Comment by kkannan1982** [comment]
I fixed this issue  with the following command 
```shell
sudo apt-get remove golang-docker-credential-helpers
```

**Comment by lallolu** [comment]
We solved it on out machine with 
`sudo apt-get install pass gnupg2`

**Comment by thaJeztah** [comment]
Yes, older versions of the `golang-docker-credential-helpers` package that's packaged by Ubuntu have this bug. That package contains all credential helpers, and docker will pick the "buggy" one, depending on what's installed on your system. Installing `pass` will make it pick the "non-buggy" one.

That said; Ubuntu ("canonical") should have the package fixed, as this is a bug that's fixed upstream, but they didn't patch some of their distro versions;


Ubuntu 20.04 LTS should have the fix

```console
$ apt-cache madison golang-docker-credential-helpers
golang-docker-credential-helpers |    0.6.3-1 | http://archive.ubuntu.com/ubuntu focal/universe amd64 Packages
```

Ubuntu 18.04 LTS has an outdated version of the package, that does not have the fix; if you're running Ubuntu 18.04, I'd recommend opening a ticket in Ubuntu's issue tracker to request either the fix to be backported, or to have the version of the package updated;

```console
$ apt-cache madison golang-docker-credential-helpers
golang-docker-credential-helpers |    0.5.0-2 | http://archive.ubuntu.com/ubuntu bionic/universe amd64 Packages
```

Ubuntu 16.04 LTS doesn't look to have that package

```console
$ apt-cache madison golang-docker-credential-helpers
N: Unable to locate package golang-docker-credential-helpers
```

**Comment by peterwillcn** [comment]
```
cd ${HOME}
wget https://github.com/docker/docker-credential-helpers/releases/download/v0.6.3/docker-credential-secretservice-v0.6.3-amd64.tar.gz
tar -xf docker-credential-secretservice-v0.6.3-amd64.tar.gz
chmod +x docker-credential-secretservice
mv /usr/bin/docker-credential-secretservice /usr/bin/docker-credential-secretservice.bkp
mv docker-credential-secretservice /usr/bin/
```
